### PR TITLE
rustup for big refactor; kill most of validation

### LINKED ISSUE #382: Fix remaining compile-fail tests
* [ ] compile-fail/match_char2.rs
* [ ] compile-fail/memleak.rs
* [ ] compile-fail/memleak_rc.rs
* [x] compile-fail/overflowing-rsh-2.rs
* [ ] compile-fail/static_memory_modification.rs

### PR DESCRIPTION
**Pull request by RalfJung**
Do not force-push, rustc refers to this commit with https://github.com/rust-lang/rust/pull/53424

Fixes #382
Fixes https://github.com/solson/miri/issues/439

### CODE REVIEW

**Comment by RalfJung** [comment]
Argh, again new new nightly :(

**Comment by RalfJung** [comment]
This now fixes the compile-fail tests for memory leaks and mutating statics. :) It also gets rid of some hacks that hook some arbitrary functions even though they have MIR... I will probably pay for this by breaking macOS and/or Windows. Let's see.

**Comment by RalfJung** [comment]
(Just experimenting with AppVeyor)

**Comment by RalfJung** [comment]
We finally have a nightly with the refactorings this matches, so I am going to try and make CI pass. Otherwise I think this is ready for review.

@oli-obk do you want be to rebase and maybe squash a little? We ended up backing out the miri updates, so I *think* none of the commits here is actually ever referenced from rustc.

The only actually functional changes in this PR are:
* Removing the unconditional hook of "std::sys::unix::thread::guard::init", "std::sys::unix::thread::guard::current" and "std::sys::unix::fast_thread_local::register_dtor".  Instead, we can now emulate foreign statics so these functions can actually properly execute.
* Hook some more `pthread_` functions that initialization needs now, and avoid hooking *all* `pthread:*` to avoid papering over unimplemented foreign items.
* Use new mechanism provided by miri engine for handling (mutable) statics.
* Re-use intrinsics and lang item implementations from CTFE where possible.

**Comment by oli-obk** [comment]
squashing/rebasing seems fine to me, but not necessary

**Comment by RalfJung** [comment]
Well then I'd say let's get this merged. CI is green. Waiting for your r+ :)

File: src/lib.rs [hunk 1]
```diff
@@ -293,15 +210,15 @@ pub struct Evaluator<'tcx> {
     /// Miri does not expose env vars from the host to the emulated program
     pub(crate) env_vars: HashMap<Vec<u8>, Pointer>,
 
-    /// Places that were suspended by the validation subsystem, and will be recovered later
-    pub(crate) suspended: HashMap<DynamicLifetime, Vec<ValidationQuery<'tcx>>>,
+    /// Use the lifetime
```
Comment (oli-obk) [inline src/lib.rs]: lolwat? Are you going to need it again in the future? Or why not just remove it? Too much churn?

Comment (RalfJung) [reply src/lib.rs]: You mean remove the lifetime? Hm, I didn't remember adding it so I just did not touch it.^^  But also indeed I am not sure yet which extra data I will add again for the Stacked Borrows stuff.

**Pull request merged**